### PR TITLE
build: create a make target for test-with-coverage

### LINKED ISSUE #461: make test: ninja: error: unknown target 'coverage-html'
## Describe the bug

**make test** works but is failing in the end with `ninja: error: unknown target 'coverage-html'`:

```
# make test 
<SNIP>
meson test -C builddir
ninja: Entering directory `/root/test/bluechi/builddir'
ninja: no work to do.
 1/11 cfg_test                     OK              0.04s
 2/11 list_test                    OK              0.04s
 3/11 parse-util_test              OK              0.03s
 4/11 string-util_test             OK              0.03s
 5/11 time-util_test               OK              0.03s
 6/11 is_ipv4_test                 OK              0.02s
 7/11 is_ipv6_test                 OK              0.02s
 8/11 get_address_test             OK              0.02s
 9/11 assemble_tcp_address         OK              0.02s
10/11 get_log_timestamp            OK              0.01s
11/11 get_time_seconds             OK              0.01s

Ok:                 11
Expected Fail:      0
Fail:               0
Unexpected Pass:    0
Skipped:            0
Timeout:            0

Full log written to /root/test/bluechi/builddir/meson-logs/testlog.txt
ninja coverage-html -C builddir
ninja: Entering directory `builddir'
ninja: error: unknown target 'coverage-html'
make: *** [Makefile:19: test] Error 1
```

### PR DESCRIPTION
**Pull request by dougsland**
Currently the make check command will try to generate a coverage report without check if any tool is available (for such task without validation can lead to an error).

This patch split the make check in three parts:

- make test: to the tests without generating the coverage report

- make test-with-coverage: execute the tests and generate the report

- make check-coverage-tools: look for required tools to be used

Fixes: #461

Signed-off-by: Douglas Schilling Landgraf <dougsland@redhat.com>

### CODE REVIEW

File: build-scripts/check-required-tools-for-coverage [hunk 1]
```diff
@@ -0,0 +1,20 @@
+#!/bin/bash
+# SPDX-License-Identifier: GPL-2.0-or-later
+
+TOOLS_COVERAGE_REPORT=("lcov" "gcovr" "genhtml")
+FOUND=0
+
+for cmd in "${TOOLS_COVERAGE_REPORT[@]}"; do
```
Comment (engelmi) [inline build-scripts/check-required-tools-for-coverage:7]: This isn't really correct. Based on https://mesonbuild.com/Unit-tests.html#coverage, it is either (`lcov` and `genhtml`) or `gcovr` - so only `lcov` or only `genhtml` isn't sufficient.

**Review by engelmi — COMMENTED** [review]
The make target split is fine since `test` does not imply generating an html coverage report. 
However, I am not convinced about the pros of the check script. Its only purpose is for checking if certain tools used in an edge case during development (in our case) are installed and providing a plain message. 
Additionally, if someone is getting this message and installed the tools correctly, the ninja target `coverage-html` would still be missing so the next error would be the error this script tries to prevent - `meson` would require a re-setup (maybe `--wipe` is enough) to generate the proper ninja targets (at least as far as I know, but haven't checked).

**Review by mwperina — COMMENTED** [review]
It doesn't seem to me worth to detect if a tool is present or not. I'd just do following changes:

1. Split test coverage generation into 2 targets: `test` and `test-coverage` (`test-coverage` would depend on test and execute coverage report afterwards)
2. Document in `README.developer.md` that additional package (gcovr or lcov) is required to run coverage report

File: build-scripts/check-required-tools-for-coverage (hunk 1 above)
Comment (dougsland) [reply build-scripts/check-required-tools-for-coverage:7]: nice! Thanks for that.

**Comment by dougsland** [comment]
> The make target split is fine since `test` does not imply generating an html coverage report.
> However, I am not convinced about the pros of the check script. Its only purpose is for checking if certain tools used in an
> edge case during development (in our case) are installed and providing a plain message.
> Additionally, if someone is getting this message and installed the tools correctly, the ninja target `coverage-html` would
> still be missing so the next error would be the error this script tries to prevent - `meson` would require a re-setup (maybe
> `--wipe` is enough) to generate the proper ninja targets (at least as far as I know, but haven't checked).

The problem is that the ninja script has a cryptography message, which doesn't help at all:

```
ninja: Entering directory `builddir'
ninja: error: unknown target 'coverage-html'
make: *** [Makefile:19: test] Error 1
```

**Comment by dougsland** [comment]
> > The make target split is fine since `test` does not imply generating an html coverage report.
> > However, I am not convinced about the pros of the check script. Its only purpose is for checking if certain tools used in an
> > edge case during development (in our case) are installed and providing a plain message.
> > Additionally, if someone is getting this message and installed the tools correctly, the ninja target `coverage-html` would
> > still be missing so the next error would be the error this script tries to prevent - `meson` would require a re-setup (maybe
> > `--wipe` is enough) to generate the proper ninja targets (at least as far as I know, but haven't checked).
> 
> The problem is that the ninja script has a cryptography message, which doesn't help at all:
> 
> ```
> ninja: Entering directory `builddir'
> ninja: error: unknown target 'coverage-html'
> make: *** [Makefile:19: test] Error 1
> ```

As soon the tools are installed, no need to generate anything in meson side, just run `make check` or `make check-with-coverage`

**Review by sandrobonazzola — COMMENTED** [review]
The idea of having a specific target with coverage was also discussed on https://github.com/containers/bluechi/pull/436 and rejected already.

**Comment by dougsland** [comment]
> The idea of having a specific target with coverage was also discussed on #436 and rejected already.

So, looks like it's  I am not only one complaining about it.
Unfortunately, others users/developers will face it too. The ninja output is nothing useful at all.
